### ISSUE TITLE
SubscriptionHandler type issue

### ISSUE THREAD

**Issue by mjadczak** [FormidableLabs/urql#319]
The `SubscriptionHandler` used for generating a composite response from subscription events is currently typed as
```typescript
export type SubscriptionHandler<T, R> = (prev: R | void, data: T) => R;
```
but the `void` type should not be used in argument types unless it is to disallow them from being used. For instance, the trivial handler
```typescript
(updates: number = 0, response) => {
      return 42
    }
```
errors out with
```
error TS2345: Argument of type '(updates: number | undefined, response: ServerImageEventsSubscription) => number' is not assignable to parameter of type 'SubscriptionHandler<ServerImageEventsSubscription, ServerImageEventsSubscription>'.
  Types of parameters 'updates' and 'prev' are incompatible.
    Type 'void | ServerImageEventsSubscription' is not assignable to type 'number | undefined'.
      Type 'void' is not assignable to type 'number | undefined'.

151     (updates: number = 0, response) => {
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I think that the intention with this handler is to pass in `undefined` on the first call, when there is no previous state, and the logic of `useSubscription` indeed does this. Therefore, I think that this type should be
```typescript
export type SubscriptionHandler<T, R> = (prev: R | undefined, data: T) => R;
```

EDIT:
It seems that there is another subtle issue here: because `useSubscription` sets `R=T` by default in its type parameters, the `SubscriptionHandler` expects the "state" to be of the same type as the subscription data, unless explicitly overriden. However, using
```typescript
useSubscription<
    ServerImageEventsSubscription,
    number
  >(
    {
      query: ServerImageEventsDocument,
    },
    (updates: number = 0, response) => {
      return 4
    }
  )
```

still gives a similar (but more sane) error
```
 error TS2345: Argument of type '(updates: number | undefined, response: ServerImageEventsSubscription) => number' is not assignable to parameter of type 'SubscriptionHandler<ServerImageEventsSubscription, number>'.
  Types of parameters 'updates' and 'prev' are incompatible.
    Type 'number | void' is not assignable to type 'number | undefined'.
      Type 'void' is not assignable to type 'number | undefined'.

152     (updates: number = 0, response) => {
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by mjadczak** [comment]
There is a similar issue in `useRequest.ts`, though not in a public type (just inside the actual code):
```typescript
const prev = useRef<void | GraphQLRequest>(undefined);
```

**Comment by parkerziegler** [comment]
Thanks for reporting @mjadczak, I can put up a fix for this tomorrow.

**Comment by parkerziegler** [comment]
Released in v1.2.0.
